server/auto_updater/dataset_updater.py

```diff
@@ -0,0 +1,292 @@
+        dir_writer = csv.DictWriter(
+            dir_file, delimiter="\t", fieldnames=["tarfile", "file_path"]
+        )
+        dir_writer.writeheader()
+
+        embed_writer = csv.DictWriter(
+            embed_file,
+            delimiter="\t",
+            fieldnames=["journal", "document"] + [f"feat_{idx}" for idx in range(300)],
+        )
+        embed_writer.writeheader()
+
+        count_writer = csv.DictWriter(
+            word_file, delimiter="\t", fieldnames=["document", "lemma", "count"]
+        )
+        count_writer.writeheader()
+
+        # Cycle through each tar file on the server
+        for tar_file in tqdm.tqdm(tar_files):
+
+            # If not xml files skip
+            if all(suffix != ".xml" for suffix in Path(tar_file).suffixes):
+                continue
+            
+            # If temp file skip
+            if Path(tar_file).suffix == ".tmp":
+                continue
+
+            # Grab the file from the tarfile
+            print(f"Requesting {pmc_open_access_url}{tar_file}....")
+            requested_file_stream = request.urlopen(f"{pmc_open_access_url}{tar_file}")
+            open_stream = tarfile.open(fileobj=requested_file_stream, mode="r|gz")
+
+            # Cycle through paper members
+            # Grab new papers and get the document vectors
+            while True:
+                try:
+
+                    pmc_paper = open_stream.next()
+
+                    if pmc_paper is None:
+                        break
+
+                    if pmc_paper.isdir():
+                        continue
+
+                    if pmc_paper.name in current_pmc_set:
+                        continue
+
+                    new_paper = open_stream.extractfile(pmc_paper)
+                    doc_vector, word_counter = generate_vector_counts(
+                        word_model, new_paper, "//abstract/sec/*|//body/sec/*"
+                    )
+
+                    if word_counter is None:
+                        continue
+
+                    dir_writer.writerow(
+                        {"tarfile": tar_file, "file_path": str(pmc_paper.name)}
+                    )
+
+                    embed_writer.writerow(
+                        {
+                            "document": Path(pmc_paper.name).stem,
+                            "journal": str(Path(pmc_paper.name).parent),
+                            **dict(
+                                zip([f"feat_{idx}" for idx in range(300)], doc_vector)
+                            ),
+                        }
+                    )
+
+                    for tok in word_counter:
+                        count_writer.writerow(
+                            {
+                                "document": Path(pmc_paper.name).stem,
+                                "lemma": tok,
+                                "count": word_counter[tok],
+                            }
+                        )
+
+                except tarfile.ReadError as e:
```

Comment:
@danich1: After doing a few more tests over the weekend, I don't think this exception should be silently swallowed. To give you an idea: this exception was thrown on my test box after processing **1,084** xml files in `comm_use.A-B.xml.tar.gz`, but this tarball file includes totally **385,048** xml files, so most of them were ignored when the `break` line on 156 went out of `while` loop. 

I also downloaded `comm_use.A-B.xml.tar.gz` onto my hard drive and run `generate_new_papers()` function on this tarball file only. It took more than 10 hours and found 63,192 new papers. You can modify the function and test it by yourself. Since the FTP site includes 8 xml tarball files and their total size is over 50GB, the step to generate new papers is going to be much longer than expected, unless we use multi-thread or multi-process to speed up. I need to do more tests on this step.

When you ran the whole process on your local computer, did you ever gotten this exception? How did did it take to finish the `generate_new_papers()` step? How many new papers did it eventually find?

Reply:
Oh man! 

>When you ran the whole process on your local computer, did you ever gotten this exception? 

Yup I did. Hence why I added the exception with a break statement at the end. I though skipping and waiting for PMCOA peeps to fix this file would be sufficient. Assumed the file was just broken.

>How did did it take to finish the generate_new_papers() step?
> How many new papers did it eventually find?

Like 30-45 minutes since the amount of new papers were tiny (1221 new papers). Given this discovery I agree multi-processing is needed for a needed speed up on this function along with: `update_paper_bins_stats()`